File: 2026-03/2026-03-22/coffee_roast_detector.py
def detect_roast(beans):
    BEAN_VALUES = {
        '\'': 1,
        '-': 2,
        '.': 3
    }
    LIGHT_TO_MEDIUM_THRESHOLD = 1.75
    MEDIUM_TO_DARK_THRESHOLD = 2.5

    roast_level = sum([BEAN_VALUES[bean] for bean in list(beans)]) / len(beans)

    return "Light" if roast_level < LIGHT_TO_MEDIUM_THRESHOLD \
        else "Medium" if roast_level <= MEDIUM_TO_DARK_THRESHOLD \
        else "Dark"

# --- TEST SUITE ---

tests_text = r"""
1. detect_roast("''-''''''-'-''--''''") should return "Light".
2. detect_roast(".'-''-''..'''.-.-''-") should return "Medium".
3. detect_roast("--.''--'-''.--..-.--") should return "Medium".
4. detect_roast("-...'-......-..-...-") should return "Dark".
5. detect_roast(".--.-..-......----.'") should return "Medium".
6. detect_roast("..-..-..-..-....-.-.") should return "Dark".
7. detect_roast("-'-''''''..-'.''-'.'") should return "Light".
"""

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 03 March/03-22/coffeeRoastDetector.js
function detectRoast(beans) {
  const BEAN_VALUES = {
    '\'': 1,
    '-': 2,
    '.': 3
  };
  const LIGHT_TO_MEDIUM_THRESHOLD = 1.75;
  const MEDIUM_TO_DARK_THRESHOLD = 2.5;

  const roastLevel = beans.split("")
    .map(bean => BEAN_VALUES[bean])
    .reduce((sum, value) => sum + value)
    / beans.length;

  return roastLevel < LIGHT_TO_MEDIUM_THRESHOLD ? "Light" 
    : roastLevel <= MEDIUM_TO_DARK_THRESHOLD ? "Medium" 
    : "Dark";
}

// --- TEST SUITE ---

const testsText = String.raw`
1. detectRoast("''-''''''-'-''--''''") should return "Light".
2. detectRoast(".'-''-''..'''.-.-''-") should return "Medium".
3. detectRoast("--.''--'-''.--..-.--") should return "Medium".
4. detectRoast("-...'-......-..-...-") should return "Dark".
5. detectRoast(".--.-..-......----.'") should return "Medium".
6. detectRoast("..-..-..-..-....-.-.") should return "Dark".
7. detectRoast("-'-''''''..-'.''-'.'") should return "Light".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);